app/controllers/guide_article_drafts_controller.ts

```diff
@@ -0,0 +1,260 @@
+      if (has !== true) {
+        const { ForbiddenException } = await import(
+          "#exceptions/http_exceptions"
+        );
+        throw new ForbiddenException();
+      }
+    }
+  }
+
+  // When linking to an existing article, require per-model permission on the original article
+  protected async storeHook(ctx: {
+    http: HttpContext;
+    request: PartialModel<typeof GuideArticleDraft>;
+  }) {
+    const { http, request } = ctx;
+    const originalId = request.originalArticleId;
+    if (originalId === null || originalId === undefined) {
+      return;
+    }
+
+    const user = http.auth.user as unknown as
+      | {
+          hasRole?: (slug: string) => Promise<boolean>;
+          hasPermission?: (
+            action: string,
+            target?: unknown,
+          ) => Promise<boolean>;
+        }
+      | undefined;
+    const isAdmin =
+      (await user?.hasRole?.("solvro_admin")) === true ||
+      (await user?.hasRole?.("admin")) === true;
+    if (isAdmin) {
+      return;
+    }
+
+    const article = await GuideArticle.find(originalId);
+    if (article === null) {
+      const { NotFoundException } = await import("#exceptions/http_exceptions");
+      throw new NotFoundException(
+        `GuideArticle with id ${originalId} not found`,
+      );
+    }
+    const allowed = await user?.hasPermission?.("update", article);
+    if (allowed !== true) {
+      const { ForbiddenException } = await import(
+        "#exceptions/http_exceptions"
+      );
+      throw new ForbiddenException();
+    }
+  }
+
+  $configureRoutes(
+    controller: LazyImport<
+      Constructor<BaseController<typeof GuideArticleDraft>>
+    >,
+  ) {
+    super.$configureRoutes(controller);
+    router
+      .post("/:id/approve", async (ctx) => {
+        const module = await controller();
+        const ControllerClass = module.default;
+        const instance = new ControllerClass();
+        return (
+          instance as unknown as {
+            approve: (c: HttpContext) => Promise<unknown>;
+          }
+        ).approve(ctx as unknown as HttpContext);
+      })
+      .as("approve");
+  }
+
+  async approve({ params, auth }: HttpContext) {
+    if (!auth.isAuthenticated) {
+      await auth.authenticate();
+    }
+    // Only solvro_admin can approve drafts
+    const isSolvroAdmin = await (
+      auth.user as unknown as { hasRole?: (slug: string) => Promise<boolean> }
+    ).hasRole?.("solvro_admin");
+    if (isSolvroAdmin !== true) {
+      const { ForbiddenException } = await import(
+        "#exceptions/http_exceptions"
+      );
+      throw new ForbiddenException();
+    }
+
+    const rawId: unknown = (params as unknown as { id?: string | number }).id;
+    let draftId: number | undefined;
+    if (typeof rawId === "string") {
+      const numeric = Number(rawId);
+      draftId = Number.isNaN(numeric) ? undefined : numeric;
+    } else if (typeof rawId === "number") {
+      draftId = rawId;
+    }
```

Comment:
you should use an autogenerated validator here (check how base controller implementations do this)